awx/api/views/__init__.py

```diff
@@ -1445,6 +1445,37 @@ class CredentialInputSourceSubList(SubListCreateAttachDetachAPIView):
     parent_key = 'target_credential'
 
 
+class CredentialTypeExternalTest(SubDetailAPIView):
+    view_name = _('External Credential Type Test')
+    model = models.CredentialType
+    serializer_class = serializers.EmptySerializer
+
+    def post(self, request, *args, **kwargs):
+        obj = self.get_object()
+        credential_id = request.data.get('credential', None)
+
+        test_inputs = {}
+        if credential_id is not None:
+            cred = models.Credential.accessible_objects(request.user, 'use_role').filter(id=credential_id).first()
+            if cred is None or cred.credential_type.kind != 'external':
+                return Response({}, status=status.HTTP_400_BAD_REQUEST)
+
+            for field_name in cred.inputs:
+                test_inputs[field_name] = cred.get_input(field_name)
+
+        for key, value in request.data.get('inputs', {}).items():
+            if value != '$encrypted$':
+                test_inputs[key] = value
+
+        try:
+            obj.plugin.backend(**test_inputs)
+            return Response({}, status=status.HTTP_202_ACCEPTED)
+        except RuntimeError as exc:
```

Comment:
This can definitely raise things _other_ than `RuntimeError` (the azure library will raise all sorts of exception specific to the Azure SDK.  Maybe we should just `except Exception as exc` and attempt to print _everything_?  This _might_ result in some client libraries show ugly errors, but at least it's probably a _clue_ as to what you got wrong.

Reply:
Yep - The motivation here was to maintaining tighter control on what messages can actually be displayed to the user.

We can extend this to all exceptions, but as you say we'd need to be comfortable with the idea of those toast messages saying _anything_, including extremely long error messages from libraries we don't control.

I can sort of go either way with this - So I think we can just switch this over to processing all exception types for now and reel it in later if we don't like it.